File: rust/1613_minimum_effort_path/minimum_effort_path.rs
/*
 * @Date: 2023-12-11
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-12-11
 * @FilePath: /algorithm/rust/1613_minimum_effort_path/minimum_effort_path.rs
 */

struct Solution;

impl Solution {
    pub fn minimum_effort_path(heights: Vec<Vec<i32>>) -> i32 {
        use std::cmp::max;
        const INF: i32 = 0x3f3f3f3f;
        let n = heights.len();
        let m = heights[0].len();
        let k = n * m;
        let mut e: Vec<Vec<(i32, i32)>> = vec![vec![]; k];
        for i in 0..n {
            for j in 0..m {
                if j != 0 {
                    e[i * m + j].push((
                        (i * m + j - 1) as i32,
                        (heights[i][j] - heights[i][j - 1]).abs(),
                    ));
                }
                if j + 1 != m {
                    e[i * m + j].push((
                        (i * m + j + 1) as i32,
                        (heights[i][j] - heights[i][j + 1]).abs(),
                    ));
                }
                if i != 0 {
                    e[i * m + j].push((
                        ((i - 1) * m + j) as i32,
                        (heights[i][j] - heights[i - 1][j]).abs(),
                    ));
                }
                if i + 1 != n {
                    e[i * m + j].push((
                        ((i + 1) * m + j) as i32,
                        (heights[i][j] - heights[i + 1][j]).abs(),
                    ));
                }
            }
        }
        let mut f = true;
        let mut dis: Vec<i32> = vec![INF; k];
        dis[0] = 0;
        while f {
            f = false;
            for i in 0..k {
                let d = dis[i];
                for &(j, w) in &e[i] {
                    if max(d, w) < dis[j as usize] {
                        dis[j as usize] = max(d, w);
                        f = true;
                    }
                }
            }
        }
        dis[k - 1]
    }
}

fn main() {
    let tests = vec![
        (vec![vec![1, 2, 2], vec![3, 8, 2], vec![5, 3, 5]], 2),
        (vec![vec![1, 2, 3], vec![3, 8, 4], vec![5, 3, 5]], 1),
        (
            vec![
                vec![1, 2, 1, 1, 1],
                vec![1, 2, 1, 2, 1],
                vec![1, 2, 1, 2, 1],
                vec![1, 2, 1, 2, 1],
                vec![1, 1, 1, 2, 1],
            ],
            0,
        ),
    ];

    for (heights, ans) in tests {
        assert_eq!(Solution::minimum_effort_path(heights), ans);
    }
}

File: cpp/1613_minimum_effort_path/minimum_effort_path.cpp
/*
 * @Date: 2023-12-11
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-12-11
 * @FilePath: /algorithm/cpp/1613_minimum_effort_path/minimum_effort_path.cpp
 */

#include <cassert>
#include <queue>
#include <tuple>
#include <vector>
using namespace std;

class Solution {
 private:
  static constexpr int dirs[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

 public:
  int minimumEffortPath(vector<vector<int>>& heights) {
    int m = heights.size();
    int n = heights[0].size();

    auto tupleCmp = [](const auto& e1, const auto& e2) {
      auto&& [x1, y1, d1] = e1;
      auto&& [x2, y2, d2] = e2;
      return d1 > d2;
    };
    priority_queue<tuple<int, int, int>, vector<tuple<int, int, int>>, decltype(tupleCmp)> q(tupleCmp);
    q.emplace(0, 0, 0);

    vector<int> dist(m * n, INT_MAX);
    dist[0] = 0;
    vector<int> seen(m * n);

    while (!q.empty()) {
      auto [x, y, d] = q.top();
      q.pop();
      int id = x * n + y;
      if (seen[id]) {
        continue;
      }
      if (x == m - 1 && y == n - 1) {
        break;
      }
      seen[id] = 1;
      for (int i = 0; i < 4; ++i) {
        int nx = x + dirs[i][0];
        int ny = y + dirs[i][1];
        if (nx >= 0 && nx < m && ny >= 0 && ny < n &&
            max(d, abs(heights[x][y] - heights[nx][ny])) < dist[nx * n + ny]) {
          dist[nx * n + ny] = max(d, abs(heights[x][y] - heights[nx][ny]));
          q.emplace(nx, ny, dist[nx * n + ny]);
        }
      }
    }

    return dist[m * n - 1];
  }
};

int main() {
  vector<tuple<vector<vector<int>>, int>> tests{
      {{{1, 2, 2}, {3, 8, 2}, {5, 3, 5}}, 2},
      {{{1, 2, 3}, {3, 8, 4}, {5, 3, 5}}, 1},
      {{{1, 2, 1, 1, 1}, {1, 2, 1, 2, 1}, {1, 2, 1, 2, 1}, {1, 2, 1, 2, 1}, {1, 1, 1, 2, 1}}, 0},
  };

  for (auto& [heights, ans] : tests) {
    assert(Solution().minimumEffortPath(heights) == ans);
  }
}